### PR TITLE
Use iterdomain graph to index non-global consumers

### PR DESCRIPTION
**Pull request by shmsong**

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/index_reference_replay.cpp [hunk 1]
```diff
@@ -528,6 +529,77 @@ class PreferredPathCompute : public IterVisitor {
     return compute.preferred_path;
   }
 };
+
+class LoopIndexingPreferredPathCompute : public IterVisitor {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/index_reference_replay.cpp:534]: Please comment this class as we'll remove the other variant that has comments.

File: torch/csrc/jit/codegen/cuda/index_idgraph_utils.cpp [hunk 2]
```diff
@@ -116,6 +154,80 @@ IndexingParameters getGlobalIndexParameters(
...
+  }
+
+  std::tie(loop_to_ind_map, zero_loops) = indexMapFromTV(
+      alloc_tv,
+      loops,
+      alloc_info.init_for_loop,
+      !index_producer,
+      double_buffer_loop);
+
+  ensureStaticIndexing(alloc_tv, alloc_info.init_for_loop, loops, alloc_id_map);
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/index_idgraph_utils.cpp:199]: Not a note for this PR, but generally a note: can't we do this pass much earlier outside of indexing? CC @naoyam

**Review by csarofeen — APPROVED** [review]
LGTM!

File: torch/csrc/jit/codegen/cuda/index_idgraph_utils.cpp (hunk 2 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/index_idgraph_utils.cpp:199]: Should be possible. All it needs to do is IterDomains that are used by any Local tensor must be statically constant, i.e., corresponding loop must be fully unrolled.

**Pull request merged**